nixos/modules/hardware/printers.nix

```diff
@@ -0,0 +1,135 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.hardware.printers;
+  ppdOptionsString = options: optionalString (options != {})
+    (concatStringsSep " "
+      (mapAttrsToList (name: value: "-o '${name}'='${value}'") options)
+    );
+  ensurePrinter = p: ''
+    ${pkgs.cups}/bin/lpadmin -p '${p.name}' -E \
+      ${optionalString (p.location != null) "-L '${p.location}'"} \
+      ${optionalString (p.description != null) "-D '${p.description}'"} \
+      -v '${p.deviceUri}' \
+      -m '${p.model}' \
+      ${ppdOptionsString p.ppdOptions}
+  '';
+  ensureDefaultPrinter = name: ''
+    ${pkgs.cups}/bin/lpoptions -d '${name}'
+  '';
+
+  # "graph but not # or /" can't be implemented as regex alone due to missing lookahead support
+  noInvalidChars = str: all (c: c != "#" && c != "/") (stringToCharacters str);
+  printerName = (types.addCheck (types.strMatching "[[:graph:]]+") noInvalidChars)
+    // { description = "printable string without spaces, # and /"; };
+
+
+in {
+  options = {
+    hardware.printers = {
+      ensureDefaultPrinter = mkOption {
```

Comment:
Can we drop the "ensure" part? It seems redundant.

Reply:
@edolstra You see a conceptual difference between `ensurePrinters` and `ensureDefaultPrinter` here, and therefore it's redundant for `defaultPrinter` but not for `ensurePrinters`, is that correct? I'm at your service. Maybe you could provide some general guidance on this, though?

I started to use those `ensure` prefixes to mark options that can only be enforced on a best effort basis on each service reload. In this concrete case, both `ensurePrinters` and `ensureDefaultPrinter` are translated into `lpadmin` command. I can't stop the user from modifying printers / change default printer, but the state will be restored. In case of `ensurePrinters`, I deliberately do not try to delete non-listed printers that the user might have added manually. Similar options in NixOS exist though which don't have that prefix.

[There's a discussion going on about that, sparked off by similar mysql / postgresql options](https://github.com/NixOS/nixpkgs/pull/55510)
I explained [my point there](https://github.com/NixOS/nixpkgs/pull/68353#issuecomment-555250666) as well.